src/cets_join.erl

```diff
@@ -129,3 +179,72 @@ apply_resolver_for_sorted(
     end;
 apply_resolver_for_sorted(LocalDump, RemoteDump, _F, _Pos, LocalAcc, RemoteAcc) ->
     {lists:reverse(LocalAcc, LocalDump), lists:reverse(RemoteAcc, RemoteDump)}.
+
+get_pids(Pid) ->
+    [Pid | cets:other_pids(Pid)].
+
+-spec check_do_not_overlap([server_pid()], [server_pid()]) -> ok.
+check_do_not_overlap(LocPids, RemPids) ->
+    LocSet = ordsets:from_list(LocPids),
+    RemSet = ordsets:from_list(RemPids),
+    case ordsets:intersection(LocSet, RemSet) of
+        [] ->
+            ok;
+        Overlap ->
+            Log = #{
+                what => check_do_not_overlap_failed,
+                local_servers => LocPids,
+                remote_servers => RemPids,
+                overlapped_servers => Overlap
+            },
+            ?LOG_ERROR(Log),
+            error(check_do_not_overlap_failed)
+    end.
+
+%% Checks that other_pids lists match for all nodes
+%% If they are not matching - the node removal process could be in progress
+-spec check_fully_connected([server_pid()]) -> ok.
+check_fully_connected(Pids) ->
+    Lists = [get_pids(Pid) || Pid <- Pids],
+    case are_fully_connected_lists([Pids | Lists]) of
+        true ->
+            check_same_join_ref(Pids);
+        false ->
+            Log = #{
+                what => check_fully_connected_failed,
+                expected_pids => Pids,
+                server_lists => Lists
+            },
+            ?LOG_ERROR(Log),
+            error(check_fully_connected_failed)
+    end.
+
+%% Check that all elements of the list match (if sorted)
+are_fully_connected_lists(Lists) ->
+    length(lists:usort([lists:sort(List) || List <- Lists])) =:= 1.
+
+%% Check if all nodes have the same join_ref
+%% If not - we don't want to continue joining
+check_same_join_ref(Pids) ->
+    Refs = [pid_to_join_ref(Pid) || Pid <- Pids],
+    case lists:usort(Refs) of
+        [_] ->
+            ok;
+        _ ->
+            Log = #{
+                what => check_same_join_ref_failed,
+                refs => lists:zip(Pids, Refs)
+            },
+            ?LOG_ERROR(Log),
+            error(check_same_join_ref_failed)
+    end.
+
+pid_to_join_ref(Pid) ->
+    #{join_ref := JoinRef} = cets:info(Pid),
+    JoinRef.
+
+-spec run_step(step(), join_opts()) -> ok.
+run_step(Step, #{step_handler := F}) ->
```

Comment:
For me this is counter-intuitive, because this function isn't actually running the step, but runs some custom injected logic instead. It's a lot of bending the code to fit the tests, and gives the impression of being overly complicated...